### DISCUSSION TITLE
problem "connecttohost", please help me....

### DISCUSSION THREAD

**Discussion by Myungman** [schreibfaul1/ESP32-audioI2S#210] [Q&A]
Hello, I'm writing because I need help. Among the "audio" functions, the "connecttoSD" and "connecttoSpeech" functions work well, and the "connecttohost" function fails no matter what address I enter.

this is source code arduinoIED...

#include "Arduino.h"
#include "WiFi.h"
#include "Audio.h"

#define I2S_DOUT      25
#define I2S_BCLK      27
#define I2S_LRC       26

Audio audio;

const char* ssid =     "KT_.....";
const char* password = "d.....";

void setup() {
    Serial.begin(115200); delay(500);
    
    Serial.print("Connecting to ");
    Serial.println(ssid);
    WiFi.begin(ssid, password);
    while (WiFi.status() != WL_CONNECTED) {
      delay(500);
      Serial.print(".");
    }
    Serial.print("\n  WiFi connected, IP address : ");
    Serial.println(WiFi.localIP());  
  
    audio.setPinout(I2S_BCLK, I2S_LRC, I2S_DOUT);
    audio.connecttohost("s1.knixx.fm/dein_webradio_64.aac"); // 64 kbp/s aac+
    //audio.connecttospeech("Nice to meet you. I am korean develeper and my name is myung man, rah.", "en");
}

void loop() {
    audio.loop();
}

void audio_info(const char *info){
    Serial.print("info        "); Serial.println(info);
}

this is error message example...

rst:0x10 (RTCWDT_RTC_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0018,len:4
load:0x3fff001c,len:1216
ho 0 tail 12 room 4
load:0x40078000,len:10944
load:0x40080400,len:6388
entry 0x400806b4
Connecting to KT_GiGA_2G_Wave2_9A45
.....
  WiFi connected, IP address : 172.30.1.13
info        PSRAM not found, inputBufferSize: 6399 bytes
info        buffers freed, free Heap: 226800 bytes
info        Connect to new host: "http://s1.knixx.fm/dein_webradio_64.aac"
info        Connect to "s1.knixx.fm" on port 80, extension "/dein_webradio_64.aac"
info        Request http://s1.knixx.fm/dein_webradio_64.aac failed!

The "connecttospeech" function is working well and works well even if you call ntptime, so Wi-Fi connection is fine.

What should I check? Should I take a closer look at the "connectohost" function?

**Comment by PeterBremer-at-Philips** [comment]
I have exactly the same issue, but I do not see any suggestion to solve it here. Anyone?

I do not have an I2S amplifier connected yet, could that cause the problem?

**Comment by PeterBremer-at-Philips** [comment]
I found a part of the solution:

Add audio.setConnectionTimeout(1000, 1000); before the audio.connecttohost

But now I get a "HTTP/1.1 502 Bad Gateway" on any URL

**Comment by schreibfaul1** [comment]
This looks like a problem with the internet access, maybe the router or firewall... ?

**Reply by PeterBremer-at-Philips** [reply]
I can not image it being a problem with the router/firewall as accessing these URLs via a normal browser on a PC via that same router/firewall is no problem at all.

**Reply by PeterBremer-at-Philips** [reply]
This is the program I have:
`
#include "Arduino.h"
#include "WiFi.h"
#include "Audio.h"

#define I2S_DOUT      26  // connect to DAC pin DIN
#define I2S_BCLK      27  // connect to DAC pin BCK
#define I2S_LRC       25  // connect to DAC pin LCK

Audio audio;

const char* ssid =     "--- secret ssid---";
const char* password = "--- secret pw---";

void setup() {
    Serial.begin(115200);
    WiFi.begin(ssid, password);
    while (WiFi.status() != WL_CONNECTED) delay(1500);
    audio.setPinout(I2S_BCLK, I2S_LRC, I2S_DOUT);
    audio.setConnectionTimeout(1000, 1000);
    if (!audio.connecttohost("http://s1.knixx.fm/dein_webradio_64.aac")) { // 64 kbp/s aac+
      Serial.println("connecttohost failed !!!");
    }
}

void loop() {
    audio.loop();
}

void audio_info(const char *info){
    Serial.print("info        "); Serial.println(info);
}
`

The output is:
info        PSRAM found, inputBufferSize: 283615 bytes
info        buffers freed, free Heap: 216600 bytes
info        Connect to new host: "http://s1.knixx.fm/dein_webradio_64.aac"
info        Connect to "s1.knixx.fm" on port 80, extension "/dein_webradio_64.aac"
info        Connection has been established in 63 ms, free Heap: 213692 bytes
info        HTTP/1.1 502 Bad Gateway
info        server: nginx/1.14.2
info        date: Wed, 18 May 2022 07:55:28 GMT
info        content-length: 173

**Comment by schreibfaul1** [comment]
https://www.makeuseof.com/what-is-the-502-error/

**Reply by PeterBremer-at-Philips** [reply]
As it works completely correct with a normal PC browser, it can not be the main cause as described in that link, a server issue. The server works perfectly fine via the same router/firewall/connection as I use for both that PC browser and the ESP32.

In that story of the link you give there is only one indication which makes for me sense in this situation:
"Browser Issues
At times, outdated browser versions ..."

**Reply by schreibfaul1** [reply]
Does the 502 error always occur with different URLs, or only with knixx.fm? It will not be a browser issue.

**Reply by PeterBremer-at-Philips** [reply]
In browser it works for:
http://s1.knixx.fm/dein_webradio_64.aac
http://vis.media-ice.musicradio.com/CapitalMP3

But I now see that the http://vis.media-ice.musicradio.com/CapitalMP3 does not give the 502 error but gives:
 MP3 decode error -2 : MAINDATA_UNDERFLOW

But http://s1.knixx.fm/dein_webradio_64.aac does give the 502 error

**Comment by schreibfaul1** [comment]
'Capital' works, MP3 decode error -2 : MAINDATA_UNDERFLOW at the beginning of a stream can occur,

'knixx.fm' doesn't work for me either, the server doesn't like 'User-Agent: ESP32 audioI2S', if you change that it works too

**Reply by PeterBremer-at-Philips** [reply]
"'Capital' works, MP3 decode error -2 : MAINDATA_UNDERFLOW at the beginning of a stream can occur"
So I understand that one is not a real issue.

"'knixx.fm' doesn't work for me either, the server doesn't like 'User-Agent: ESP32 audioI2S'"
What do you suggest? Mozilla/5.0 ?

**Reply by schreibfaul1** [reply]
yes, you can enter 'Mozilla'

**Reply by PeterBremer-at-Philips** [reply]
Thanks!!!
Will try it

**Reply by PeterBremer-at-Philips** [reply]
YES, now it works! 

Line 521 audio.cpp is now: strcat(resp, "User-Agent: Mozilla5.0\r\n");
And do not forget:
  audio.setConnectionTimeout(1000, 1000);
before:
  audio.connecttohost(...);

Great, thank you very much for the cooperation.

info        PSRAM found, inputBufferSize: 283615 bytes
info        buffers freed, free Heap: 216600 bytes
info        Connect to new host: "http://s1.knixx.fm/dein_webradio_64.aac"
info        Connect to "s1.knixx.fm" on port 80, extension "/dein_webradio_64.aac"
info        Connection has been established in 59 ms, free Heap: 213720 bytes
info        HTTP/1.1 200 OK
info        server: nginx/1.14.2
info        date: Wed, 18 May 2022 12:07:31 GMT
info        content-type: audio/aac, format is aac
info        AACDecoder has been initialized, free Heap: 188716 bytes
info        ice-audio-info: channels=2;samplerate=44100;bitrate=64
info        icy-name: knixx.fm - Dein Webradio. / 64 kbp/s aac+
info        access-control-allow-origin: *
info        access-control-allow-headers: Origin, Accept, X-Requested-With, Content-Type
info        access-control-allow-methods: GET, OPTIONS, HEAD
info        x-frame-options: SAMEORIGIN
info        x-content-type-options: nosniff
info        Switch to DATA, metaint is 16000
info        stream ready
info        buffer filled in 5 ms
info        syncword found at pos 0
info        Channels: 1
info        SampleRate: 22050
info        BitsPerSample: 16
info        BitRate: 62360
info        AAC HeaderFormat: ADTS
info        AAC Codec: MPEG-4
info        AAC Profile: LowComplexity
info        VBR recognized, audioFileDuration is estimated
info        StreamTitle='Cyndi Lauper - She Bop -- 1983'
